src/auth/permissions/index.ts

```diff
@@ -0,0 +1,173 @@
+    .then((res) => res[0].access_role_id);
+
+const permissionsInitialiser: PermissionsInitialiser = (config) => {
+  const client = knex(config);
+  return {
+    grantExisting: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert(
+              client.select({
+                ['access_role_id']: roleId,
+                ['permission_id']: permissionId,
+              })
+                .whereNotExists(client('access_role_permission').where({
+                  ['access_role_id']: roleId,
+                  ['permission_id']: permissionId,
+                }))
+            );
+        if (insertPermission.length === 0) {
+          throw new Error('Permission entry is already associated to this role');
+        }
+        client.destroy();
+        return insertPermission;
+      } else {
+        client.destroy();
+        throw new Error('Permission entry does not exists, please use grantNew method');
+      }
+    },
+    grantNew: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 0) {
+        const createPermissionId = await client('permission')
+            .returning('permission_id')
+            .insert({
+              ['permission_entity']: resource,
+              ['permission_level']: permissionLevel,
+              ['access_type']: access,
+            });
+
+        const permissionId = createPermissionId[0];
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert({
+              ['access_role_id']: roleId,
+              ['permission_id']: permissionId,
+            });
+        client.destroy();
+        return insertPermission;
+      } else {
+        throw new Error('Permission already exists, please use grantExisting method');
+        client.destroy();
+      }
+    },
+    revoke: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const deleteRow = await client('access_role_permission')
+            .where({
+              ['access_role_id']: roleId,
```

Comment:
Throughout all these methods, since `getRoleId` and `getPermissionIds` just return `knex.QueryBuilder` instances, you don't need to `await` them. You can just combine them all into a single query that uses sub-queries:
```js
client('access_role_permission')
  .del()
  .where({ 
    access_role_id: getRoleId({ client, role }),
    permission_id: getPermissionIds({ ... }),
  });
```
This way, `knex` should also error when either sub-query returns more than one row, (possibly) saving you from doing checks on outputs.

Reply:
From what I can see, sql doesnt error if subqueries/where statements return nothing in delete statements. This means I can't return appropriate errors if this is done in one query.